beancount/plugins/close_tree.py

```diff
@@ -0,0 +1,98 @@
+"""This plugin inserts close directives for all of an account's descendants when an account is
+closed. Unopened parent accounts can also be closed. Any explicitly specified close is left
+untouched.
+
+For example, given this:
+
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage  ; this account does not necessarily need to be opened
+```
+
+the plugin turns it into:
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage:AAPL
+2018-11-10 close Assets:Brokerage:ORNG
+```
+
+Invoke this plugin _after_ any plugins that generate `open` directives for account trees that you
+want to auto close. An example is the `auto_accounts` plugin that ships with Beancount:
+
+```
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
+
+import time
+from beancount.core import data
+from beancount.core.data import Open, Close
+
+DEBUG = 0
+__plugins__ = ('close_tree',)
+
+
+def close_tree(entries, unused_options_map):
+    """Insert close entries for all subaccounts of a closed account.
+
+    Args:
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
+
+    start_time = time.time()
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
```

Comment:
If you wanted to be more precise about closing really just the subaccounts that are open at the time of closing, you should keep track of the tree of open accounts as you process the directives and close the subaccounts that are open then. You'd also have to handle subsequent closing directives colliding with your inserted ones. But I think that would be the most correct algorithm.

Reply:
To be more clear what I mean is that it's possible for accounts to be opened, closed and then reopened (I think, I forget a bit).

Reply:
```
2014-01-01 open Assets:XBank
2015-01-01 close Assets:XBank

2016-01-01 open Assets:XBank
2017-01-01 close Assets:XBank   
```

This doesn't seem to be supported (in v2, at least). I get:
```
Duplicate open directive for Assets:XBank
Duplicate close directive for Assets:XBank
```

Do we need to address this case?

Reply:
Ah, found the [documentation](https://beancount.github.io/docs/beancount_language_syntax.html#directives):
At the moment, once an account is closed, you cannot reopen it after that date.